### DISCUSSION TITLE
Meson build system support?

### DISCUSSION THREAD

**Discussion by ethindp** [pocoproject/poco#4706] [Ideas]
**Is your feature request related to a problem? Please describe.**
I'd like to use Poco with Meson

**Describe the solution you'd like**
It would be cool if Meson support was added. CMake can't see Meson wrap-based dependencies at the moment.

**Describe alternatives you've considered**
I'm uncertain what alternatives there are. If one just uses the basic edition you lose out on TLS/SSL support and things like that on from what I can tell pretty much any platform.

**Additional context**
I'm hoping to migrate a project away from scons to Meson, hence this request.

**Comment by yoboujon** [comment]
Hey ! I was wondering how to make it work with [Meson](https://mesonbuild.com/) like you, and I tried to make a configuration file that seem to work. Just note that there is an issue for now and I think it may be best to create a fully-fledged `meson.build` that focuses on Poco.

## Simple configuration files

Here is what you should add to your `meson.build`:
```meson
cmake = import('cmake')
poco_proj = cmake.subproject('poco')
poco_dep_foundation = poco_proj.dependency(['Foundation'])
poco_dep_util = poco_proj.dependency(['Util'])

# add to your executable 'dependencies: [poco_dep_foundation, poco_dep_util]' afterwards
```

Then to your `subprojects/` create the file `poco.wrap`:
```yaml
[wrap-file]
directory = poco-poco-1.14.1-release
source_url = https://github.com/pocoproject/poco/archive/refs/tags/poco-1.14.1-release.tar.gz
source_filename = poco-1.14.1-release.tar.gz
source_hash = e711712d68d01e580be315e1cca21d2d10ce941fc3252c0402c0a5732f719445
method = cmake
wrapdb_version = 1.14.1
```
_Here it is for the version [1.14.1](https://github.com/pocoproject/poco/tree/poco-1.14.1-release), but it can work with other versions as well, just make sure to update the source hash_

## What are the issues for now ?

### Library name

It seem to work well on **Linux** because the bindings are done on the go to the shared/static library. However, it seems that the name generated from Meson is not the correct one. I get `libFoundation.so` instead of `libPocoFoundation.so` this is because in Meson, the generated library name is gathered from the **target CMake name** and not the **OUTPUT_NAME**, which if you look closely in the CMakeLists.txt from Foundation is:

```cmake
set_target_properties(Foundation
	PROPERTIES
	VERSION ${SHARED_LIBRARY_VERSION} SOVERSION ${SHARED_LIBRARY_VERSION}
	OUTPUT_NAME PocoFoundation
	DEFINE_SYMBOL Foundation_EXPORTS
)
```

_One workaround would be to create a fork and to modify explicitly these lines. However this cannot be a definitive solution as you guessed._

### Generating `pocomsg.h`

Another issue, on Windows especially, is the build order. Poco generates a file `pocomsg.h` which is used afterwards in Foundation, however it seems that Meson doesn't build the targets in order and therefore does not generate the header which seems to be missing during compilation.
```
FAILED: subprojects/poco-poco-1.9.4-release/Foundation.dll.p/Foundation_src_EventLogChannel.cpp.obj
"cl" "-Isubprojects\poco-poco-1.9.4-release\Foundation.dll.p" "-Isubprojects\poco-poco-1.9.4-release" "-I..\subprojects\poco-poco-1.9.4-release" "-Isubprojects\poco-poco-1.9.4-release\__CMake_build\Foundation" "-I..\subprojects\poco-poco-1.9.4-release\__CMake_build\Foundation" "-I..\subprojects\poco-poco-1.9.4-release\Foundation\include" "-I..\subprojects\poco-poco-1.9.4-release\Foundation\src" "-Isubprojects\poco-poco-1.9.4-release\__CMake_build" "-I..\subprojects\poco-poco-1.9.4-release\__CMake_build" "/MDd" "/nologo" "/showIncludes" "/utf-8" "/Zc:__cplusplus" "/EHsc" "/std:c++14" "/permissive-" "/Od" "/Zi" "/DWIN32" "/D_WINDOWS" "/GR" "/EHsc" "/MDd" "-DFoundation_EXPORTS" "-DPCRE_STATIC" "-DPOCO_ENABLE_CPP11" "-DPOCO_ENABLE_CPP14" "-DPOCO_NO_AUTOMATIC_LIBS" "-DPOCO_OS_FAMILY_WINDOWS" "-DUNICODE" "-D_UNICODE" "-D__LCC__" "/Fdsubprojects\poco-poco-1.9.4-release\Foundation.dll.p\Foundation_src_EventLogChannel.cpp.pdb" /Fosubprojects/poco-poco-1.9.4-release/Foundation.dll.p/Foundation_src_EventLogChannel.cpp.obj "/c" ../subprojects/poco-poco-1.9.4-release/Foundation/src/EventLogChannel.cpp
../subprojects/poco-poco-1.9.4-release/Foundation/src/EventLogChannel.cpp(18): fatal error C1083: Impossible d'ouvrir le fichier include : 'pocomsg.h' : No such file or directory
```

## Implications

Because the generated name of the library is not the one expected by the underlying CMake tool-chain, this leads to either:
- **Windows**: the build failed because other components from Poco library cannot link to each other (tries to find PocoFoundation.dll but it has been generated as Foundation.dll)
- **Windows**: because the header cannot be generated, the source code cannot be generated afterwards
- **Linux**: if using shared libraries, the name linked will not correspond and therefore your program could not be run outside of your environment.
